module/zfs/vdev_removal.c

```diff
@@ -2038,7 +2046,20 @@ spa_vdev_remove_top_check(vdev_t *vd)
 	int num_indirect = 0;
 	for (uint64_t id = 0; id < rvd->vdev_children; id++) {
 		vdev_t *cvd = rvd->vdev_child[id];
-		if (cvd->vdev_ashift != 0 && !cvd->vdev_islog)
+
+		/*
+		 * A removed special/dedup vdev must have the same ashift
+		 * across all vdevs in its class.
+		 */
+		if (cvd->vdev_alloc_bias != VDEV_BIAS_NONE &&
```

Comment:
I think it's more clear to say that we only care about this when removing a special/dedup vdev (even though the logic is equivalent since we also only care if the child vdev has the same alloc bias as the device being removed).
```suggestion
		if (vd->vdev_alloc_bias != VDEV_BIAS_NONE &&
```